infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/runtime/fsm/sink/VerifyRemoteSourceLeader.java

```diff
@@ -0,0 +1,238 @@
+import java.util.concurrent.TimeUnit;
+
+import static org.corfudb.protocols.CorfuProtocolCommon.getUuidMsg;
+import static org.corfudb.protocols.service.CorfuProtocolMessage.getDefaultProtocolVersionMsg;
+import static org.corfudb.protocols.service.CorfuProtocolMessage.getResponseMsg;
+
+@Slf4j
+public class VerifyRemoteSourceLeader {
+
+    /**
+     * Executor service for FSM event queue consume
+     */
+    private final ExecutorService communicationFSMConsumer;
+
+    /**
+     * Remote nodes to which connection has been established.
+     */
+    private final Set<String> connectedNodes;
+
+    /**
+     * A queue of events.
+     */
+    private final LinkedBlockingQueue<LogReplicationRuntimeEvent> eventQueue = new LinkedBlockingQueue<>();
+
+    /**
+     * Session information
+     */
+    private final LogReplicationSession session;
+
+    /**
+     * A router that forwards a message to the transport layer
+     */
+    private final LogReplicationClientServerRouter router;
+
+    /**
+     * NodeId of the remote leader.
+     */
+    private volatile Optional<String> leaderNodeId = Optional.empty();
+
+    public VerifyRemoteSourceLeader(LogReplicationSession session, LogReplicationClientServerRouter router) {
+        this.session = session;
+        this.router = router;
+        this.connectedNodes = new HashSet<>();
+
+        this.communicationFSMConsumer = Executors.newSingleThreadExecutor(new
+                ThreadFactoryBuilder().setNameFormat(
+                "sink-consumer-"+session.hashCode()).build());
+
+
+        communicationFSMConsumer.submit(this::consume);
+    }
+
+    public synchronized void input(LogReplicationRuntimeEvent event) {
+        try {
+            log.info("adding to the queue {}", event);
+            eventQueue.put(event);
+        } catch (InterruptedException ex) {
+            log.error("Log Replication interrupted Exception: ", ex);
+        }
+    }
+
+    /**
+     * Consumer of the eventQueue.
+     * <p>
+     * This method consumes the log replication events and does the state transition.
+     */
+    private void consume() {
+        try {
+            //  Block until an event shows up in the queue.
+            LogReplicationRuntimeEvent event = eventQueue.take();
+            processEvent(event);
+
+            communicationFSMConsumer.submit(this::consume);
+
+        } catch (Throwable t) {
+            log.error("Error on event consumer: ", t);
+        }
+    }
+
+    private void processEvent(LogReplicationRuntimeEvent event) {
+        log.info("processing event {}", event);
+        switch (event.getType()) {
+            case ON_CONNECTION_DOWN:
+                String nodeIdDown = event.getNodeId();
+                log.info("Detected connection down from node={}", nodeIdDown);
+                updateDisconnectedNodes(nodeIdDown);
+                resetRemoteLeader();
+                break;
+            case REMOTE_LEADER_NOT_FOUND:
```

Comment:
Just throw out some thoughts not sure will it be cleaner or not, by reusing the `LogReplicationRuntimeEventType` enum we sort of bring these events to Sink side, whereas `LogReplicationRuntime` is actually dedicated for Source side.

do you think it worth it to create events just for this sink side eventQueue/stateMachine under `fsm/sink/`?